Add docker support

Firefishy opened the pull request:
* Lock required python packages
* Add Docker support
* Add GHA to build and publish docker container to github (arm64 + amd64)

Firefishy committed:
Lock python requirements.txt

Firefishy committed:
Add Docker support

Signed-off-by: Grant Slater <github@firefishy.com>

tomhughes commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+ARG CONFD_VERSION=0.19.1
+RUN CONFD_ARCH=$(arch | sed s/aarch64/arm64/ | sed s/x86_64/amd64/) \
+    && curl -SL "https://github.com/abtreece/confd/releases/download/v${CONFD_VERSION}/confd-v${CONFD_VERSION}-linux-${CONFD_ARCH}.tar.gz" | tar -xz -C /usr/local/bin/ \
+    && /usr/local/bin/confd -version
```
Is there seriously nothing packaged we can use for this?

Firefishy replied:
I've previously used `confd` because of it's flexibility and I quite like it. Unfortunately`confd` isn't packaged in Ubuntu or Debian. The install is complicated by adding in the ARM64 support.

Using `envsubst` from gettext could likely work as a substitute. Alternatively I could likely just use python's ability to read directly environment variables.

Firefishy replied:
Replaced with simplified `gettext envsubst` using distro package. 👍

tomhughes commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+    && curl -SL "https://github.com/abtreece/confd/releases/download/v${CONFD_VERSION}/confd-v${CONFD_VERSION}-linux-${CONFD_ARCH}.tar.gz" | tar -xz -C /usr/local/bin/ \
+    && /usr/local/bin/confd -version
+
+RUN groupadd --system --gid 510 osqa && useradd --system --gid osqa --uid 510 osqa
```
Does a specific UID/GID matter once we're in a container?

Firefishy replied:
It is still recommended not to run apps as root within a root container. 

The UID / GID will be for compatibility later with existing OSQA user uploads when I enable a docker volume mount to store the uploads outside the container.

tomhughes replied:
I wasn't suggesting using root inside the container, just letting useradd pick a random ID but sharing with the host is indeed a good reason to fix it.

tomhughes commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+RUN groupadd --system --gid 510 osqa && useradd --system --gid osqa --uid 510 osqa
+
+WORKDIR /usr/src/app
+RUN chown osqa:osqa /usr/src/app
```
Why `/usr/src` here? It's likely doubly wrong because `/usr` is for the OS and package manager and `/usr/src` if it's anything is historically the OS source directory...

Firefishy replied:
Moved it to `/srv/app` the location is fairly arbitrary. The container itself is disposable.

Firefishy replied:
I've also now set ARGS (build time only environment variables) which can be overridden if desired to set preferred container UID/GID.

tomhughes commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+COPY requirements.txt ./
+RUN pip install --no-cache-dir -r requirements.txt
+
+COPY --chown=osqa . .
```
Why are we copying the current directory to itself? Is this just a long winded way of writing `chown` or something? How did any files get into this directory anyway?

Firefishy replied:
This is to `COPY` the files from current working directory (outside the container, `.` context is the Dockerfile path) into the container's `WORKDIR`.

The `--chown` sets the owner within the container to allow python to create `.pyc` cache files.

tomhughes commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+ENV OSQA_DB_ENGINE=django.db.backends.postgresql_psycopg2
+ENV OSQA_DB_NAME=osqa
+ENV OSQA_DB_USER=osqa
+ENV OSQA_DB_PASSWORD=osqa
```
If you're going to use network connections and password authentication for the database then this needs to be a secure password and not be exposed in a public repository.

Firefishy replied:
These are only the "default" credentials and will not use them in production.
We will not use the `docker-compose.yml` file in production

tomhughes commented on docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+        condition: service_started
+
+  postgres:
+    image: postgres:10
```
Why postgres 10 when we currently use 14 and 15 is now available?

Firefishy replied:
I have gone for what we are currently using in production. I am unsure about OSQA's postgres compatibility.

tomhughes replied:
You haven't though - as I said we are using 14 in production!

Firefishy replied:
shenron is running `postgresql-10 10.22-1.pgdg20.04+1`

```
/usr/lib/postgresql/10/bin/postgres -D /var/lib/postgresql/10/main -c config_file=/etc/postgresql/10/main/postgresql.conf
```

Firefishy replied:
Only the client tools are postgres `14`

tomhughes replied:
Damn you're right... I could have sworn I had upgraded all the old postgres instances but I see that one is not being monitored in prometheus which is why I missed it so that's two things to fix!

Firefishy replied:
Updated to postgres 14 now.

tomhughes commented on docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+    image: postgres:10
+    environment:
+      POSTGRES_USER: osqa
+      POSTGRES_PASSWORD: osqa
```
See above...

Firefishy replied:
Only used for testing locally. `docker-compose.yml` will not be used in production.

tomhughes commented on docker-compose.yml:
```diff
@@ -0,0 +1,35 @@
+      postgres:
+        condition: service_healthy
+      memcached:
+        condition: service_started
```
Do we want to run these in separate containers? I think for discourse we put the database on the host didn't we?

Firefishy replied:
Docker the best practise is one task one container. The `docker-compose.yml` is only an example used locally for testing.

tomhughes commented on Dockerfile:
```diff
@@ -0,0 +1,52 @@
+# hadolint global ignore=DL3008
+FROM debian:buster
```
Why debian rather than ubuntu?

Firefishy replied:
Most docker official images (eg: python, memcached, etc) use Debian or Alpine for their base images.

Debian buster still supports python 2.7 (default python version). I am planning to test using the system packaged version of `gunicorn` as the webserver.

Only ubuntu 18.04 (no newer) has a system packaged version of `gunicorn` with python 2.7 support.

Debian buster is supported until [June 2024](https://wiki.debian.org/LTS)
Ubuntu 18.04 is supported until [April 2023](https://wiki.ubuntu.com/Releases)

tomhughes commented:
I have to say I wasn't really expecting the container definition to be in the OSQA repository. I was expecting it to be something separate...

Also I guess "choose container system" in https://github.com/openstreetmap/operations/issues/807 is resolved in favour of docker?

Firefishy commented:
> I have to say I wasn't really expecting the container definition to be in the OSQA repository. I was expecting it to be something separate...

OSQA is dead. Easiest just to "hijack" the codebase (with our mods) and add the additional functionality for adding Docker. Adding git into a docker to checkout a repo to me feels like an antipattern.
 
> Also I guess "choose container system" in [openstreetmap/operations#807](https://github.com/openstreetmap/operations/issues/807) is resolved in favour of docker?

Commented on [openstreetmap/operations#807](https://github.com/openstreetmap/operations/issues/807) ticket with options for "container system". For now plain old docker + systemd + chef is good enough.

Firefishy commented:
Updates:
* Removed `confd` and replaced with `gettext envsubst`
* Better comments for `Dockerfile`
* Updated to `docker-compose.yml` postgres 14
* Allow override of default UID/GID at built time by using ARGS (build time only environment variables)
